cmd/charm/charmcmd/plugin.go

```diff
@@ -97,116 +98,235 @@ func pluginHelpTopic() string {
+
 	// Create a channel with enough backing for each plugin.
 	description := make(chan pluginDescription, len(plugins))
 	help := make(chan pluginDescription, len(plugins))
 
 	// Exec the --description and --help commands.
 	for _, plugin := range plugins {
-		go func(plugin string) {
+		fi := plugin
+		go func() {
 			result := pluginDescription{
-				name: plugin,
+				Name:    fi.name,
+				ModTime: fi.mtime,
 			}
 			defer func() {
 				description <- result
 			}()
-			desccmd := exec.Command(plugin, "--description")
+			desccmd := exec.Command(fi.name, "--description")
 			output, err := desccmd.CombinedOutput()
 
 			if err == nil {
 				// Trim to only get the first line.
-				result.description = strings.SplitN(string(output), "\n", 2)[0]
+				result.Description = strings.SplitN(string(output), "\n", 2)[0]
 			} else {
-				result.description = fmt.Sprintf("error occurred running '%s --description'", plugin)
-				logger.Debugf("'%s --description': %s", plugin, err)
+				result.Description = fmt.Sprintf("error occurred running '%s --description'", fi.name)
+				logger.Debugf("'%s --description': %s", fi.name, err)
 			}
-		}(plugin)
-		go func(plugin string) {
+		}()
+		go func() {
 			result := pluginDescription{
-				name: plugin,
+				Name: fi.name,
 			}
 			defer func() {
 				help <- result
 			}()
-			helpcmd := exec.Command(plugin, "--help")
+			helpcmd := exec.Command(fi.name, "--help")
 			output, err := helpcmd.CombinedOutput()
 			if err == nil {
-				result.doc = string(output)
+				result.Doc = string(output)
 			} else {
-				result.doc = fmt.Sprintf("error occured running '%s --help'", plugin)
-				logger.Debugf("'%s --help': %s", plugin, err)
+				result.Doc = fmt.Sprintf("error occured running '%s --help'", fi.name)
+				logger.Debugf("'%s --help': %s", fi.name, err)
 			}
-		}(plugin)
+		}()
 	}
 	resultDescriptionMap := map[string]pluginDescription{}
 	resultHelpMap := map[string]pluginDescription{}
 	// Gather the results at the end.
 	for _ = range plugins {
 		result := <-description
-		resultDescriptionMap[result.name] = result
+		resultDescriptionMap[result.Name] = result
 		helpResult := <-help
-		resultHelpMap[helpResult.name] = helpResult
+		resultHelpMap[helpResult.Name] = helpResult
 	}
 	// plugins array is already sorted, use this to get the results in order.
 	for _, plugin := range plugins {
 		// Strip the 'charm-' off the start of the plugin name in the results.
-		result := resultDescriptionMap[plugin]
-		result.name = result.name[len(pluginPrefix):]
-		result.doc = resultHelpMap[plugin].doc
+		result := resultDescriptionMap[plugin.name]
+		result.Name = result.Name[len(pluginPrefix):]
+		result.Doc = resultHelpMap[plugin.name].Doc
 		results = append(results, result)
 	}
 	pluginDescriptionsResults = results
+
+	writeResultsCache(pluginCache, results)
 	return results
 }
 
+func readAndReturnCacheIfValid(pluginCache string, plugins []fileInfo, pluginExists map[string]int) []pluginDescription {
```

Comment:
It seems to me that this might be simpler if there were a file per plugin, you could then only re-get the changed ones.